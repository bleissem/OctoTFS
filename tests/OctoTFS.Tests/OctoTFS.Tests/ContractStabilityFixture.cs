using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using Newtonsoft.Json;
using NUnit.Framework;

namespace OctoTFS.Tests
{
    [TestFixture]
    [UseReporter(typeof (DiffReporter))]
    public class ContractStabilityFixture
    {
        [Test]
        [Description("This test helps to make sure we don't accidentally change the name nor data type of inputs people have stored in VSTS - otherwise they will likely lose data.")]
        public void EnsureInputNamesAndTypesHaveNotChanged()
        {
            var taskJsonResourceNames = Assembly.GetAssembly(GetType())
                .GetManifestResourceNames()
                .Where(name => name.EndsWith("task.json"))
                .OrderBy(name => name)
                .ToArray();

            var toApprove = taskJsonResourceNames.Aggregate(new StringBuilder(), (builder, resourceName) =>
            {
                builder.AppendIndentedLine(resourceName);

                using (var resourceStream = Assembly.GetAssembly(GetType()).GetManifestResourceStream(resourceName))
                using (var textReader = new StreamReader(resourceStream))
                using (var jsonReader = new JsonTextReader(textReader))
                {
                    dynamic task = JsonSerializer.CreateDefault().Deserialize(jsonReader);
                    foreach (dynamic input in task.inputs)
                    {
                        builder.AppendIndentedLine($"'{input.name}': {input.type}", 1);
                    }
                }

                return builder;
            }, builder => builder.ToString());

            Approvals.Verify(toApprove);
        }
    }

    public static class StringBuilderExtensions
    {
        public static StringBuilder AppendIndentedLine(this StringBuilder builder, string append, int indent = 0)
        {
            return builder.Append(new string('\t', indent)).AppendLine(append);
        }
    }
}
