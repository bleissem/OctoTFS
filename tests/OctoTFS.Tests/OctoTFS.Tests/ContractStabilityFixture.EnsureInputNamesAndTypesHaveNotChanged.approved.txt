OctoTFS.Tests.resources.VSTSExtensions.OctopusBuildAndReleaseTasks.Tasks.CreateOctopusRelease.task.json
	'ConnectedServiceName': connectedService:Generic
	'ProjectName': string
	'ReleaseNumber': string
	'Channel': string
	'ChangesetCommentReleaseNotes': boolean
	'WorkItemReleaseNotes': boolean
	'CustomReleaseNotes': string
	'DeployToEnvironment': string
	'DeployForTenants': string
	'DeployForTenantTags': string
	'DeploymentProgress': boolean
	'AdditionalArguments': string
OctoTFS.Tests.resources.VSTSExtensions.OctopusBuildAndReleaseTasks.Tasks.Deploy.task.json
	'ConnectedServiceName': connectedService:Generic
	'Project': string
	'ReleaseNumber': string
	'Environments': string
	'ShowProgress': boolean
	'DeployForTenants': string
	'DeployForTenantTags': string
	'AdditionalArguments': string
OctoTFS.Tests.resources.VSTSExtensions.OctopusBuildAndReleaseTasks.Tasks.Pack.task.json
	'PackageId': string
	'PackageFormat': string
	'PackageVersion': string
	'SourcePath': filePath
	'OutputPath': filePath
	'NuGetAuthor': string
	'NuGetTitle': string
	'NuGetDescription': string
	'NuGetReleaseNotes': string
	'NuGetReleaseNotesFile': filePath
	'Include': multiLine
	'Overwrite': boolean
OctoTFS.Tests.resources.VSTSExtensions.OctopusBuildAndReleaseTasks.Tasks.Promote.task.json
	'ConnectedServiceName': connectedService:Generic
	'Project': string
	'From': string
	'To': string
	'ShowProgress': boolean
	'DeployForTenants': string
	'DeployForTenantTags': string
	'AdditionalArguments': string
OctoTFS.Tests.resources.VSTSExtensions.OctopusBuildAndReleaseTasks.Tasks.Push.task.json
	'ConnectedServiceName': connectedService:Generic
	'Package': multiLine
	'Replace': boolean
	'AdditionalArguments': string
